### PR TITLE
[core][autoscaler][v2] scheduler - enforce min worker impl [2/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR adds partial implementation of `ResourceDemandScheduler` for min nodes enforcement. 

See https://github.com/ray-project/ray/pull/39851 for a draft prototype. 

PRs to come:
- implementation for normal resources requests [3/x]
- implementation for cluster constraints [4/x]
- impelmentation for gang scheduling [5/x]
- porting of existing tests and new tests [6/x]

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/scheduler.py [hunk 1]
```diff
@@ -242,3 +245,265 @@ def __repr__(self) -> str:
...
+
+        # 3. Schedule the gang resource requests.
+        infeasible_gang_requests = self._sched_gang_resource_requests(
+            request.gang_resource_requests
+        )
+
+        # 4. Schedule the tasks/actor resource requests
+        infeasible_requests = self._sched_resource_requests(
+            request.resource_requests,
+        )
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/scheduler.py:405]: in future PRs.

**Comment by rickyyx** [comment]
Adding unit tests to this PR.

File: python/ray/autoscaler/v2/scheduler.py [hunk 2]
```diff
@@ -242,3 +245,265 @@ def __repr__(self) -> str:
...
+    @dataclass
+    class ScheduleContext:
+        """
+        Encapsulates the context for processing one scheduling requests.
+
+        This exposes functions to read and write the scheduling nodes, to prevent
+        accidental modification of the internal state.
+        """
+
+        # The current schedulable nodes (including pending nodes and pending requests).
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:273]: yes, "pending" here includes 
1. QUEUED
2. REQUESTED
3. ALLOCATED 
4. RAY_INSTALLING

where the running ones include:
1. RAY_RUNNING

File: python/ray/autoscaler/v2/scheduler.py [hunk 3]
```diff
@@ -242,3 +247,278 @@ def __repr__(self) -> str:
...
+
+    def _init_context(self, request: SchedulingRequest) -> None:
+        self._ctx = self.ScheduleContext.from_schedule_request(request)
+
+    def _enforce_min_workers(self) -> None:
+        """
+        Enforce the minimal count of nodes for each worker node type.
+
+        Args:
+            config: The cluster config.
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:440]: No such arg

File: python/ray/autoscaler/v2/scheduler.py [hunk 4]
```diff
@@ -242,3 +247,278 @@ def __repr__(self) -> str:
...
+                new_nodes.extend(
+                    [
+                        SchedulingNode.from_node_config(
+                            copy.deepcopy(node_type_config),
+                            status=SchedulingNodeStatus.TO_LAUNCH,
+                        )
+                    ]
+                    * (min_count - cur_count)
+                )
+        # NOTE: we assume the aggregated number of min workers across all node types
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:466]: Do we have validation somewhere to ensure this?

**Review by jjyao — APPROVED** [review]

File: python/ray/autoscaler/v2/scheduler.py (hunk 4 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:466]: this should be something our config checks.

**Pull request merged**